jsrc/conversions.cpp

```diff
@@ -0,0 +1,637 @@
+    auto const infinity = [](auto p) { return p < -2 || 2 < p; };
+    return convert<D, bool>(jt, w, yv, [&](auto p) {
+        return value_if(!infinity(p) && (p == 0.0 || fuzzy_equal(p, 1.0, fuzz)), p != 0.0);
+    });
+}
+
+template <>
+[[nodiscard]] auto
+convert<D, I>(J jt, array w, void *yv, D fuzz) -> bool {
+    return convert<D, I>(jt, w, yv, [&](auto p) -> std::optional<I> {
+        auto const q = jround(p);
+        if (!(p == q || fuzzy_equal(p, q, fuzz))) {
+            return std::nullopt;  // must equal int, possibly out of range
+        }
+        // out-of-range values don't convert, handle separately
+        if (p < static_cast<D> IMIN) {
+            return value_if(p >= IMIN * (1 + fuzz), IMIN);
+        }  // if tolerantly < IMIN, error; else take IMIN
+        else if (p >= FLIMAX) {
+            return value_if(p <= -static_cast<D> IMIN * (1 + fuzz), IMAX);
+        }  // if tolerantly > IMAX, error; else take IMAX
+        return q;
+    });
+}
+
+template <>
+[[nodiscard]] auto
+convert<Z, D>(J jt, array w, void *yv, D fuzz) -> bool {
+    if (fuzz != 0.0) {
+        return convert<Z, D>(jt, w, yv, [&](auto const &v) {
+            auto const d = std::abs(v.im);
+            return value_if(d != inf && d <= fuzz * std::abs(v.re), v.re);
+        });
+    }
+    return convert<Z, D>(jt, w, yv, [](auto v) { return value_if(!v.im, v.re); });
+}
+
+template <>
+[[nodiscard]] auto
+convert<bool, X>(J jt, array w, void *yv) -> bool {
+    return convert<B, X>(jt,
+                         w,
+                         yv,
+                         [=](auto v) {
+                             int64_t u[] = {v};
+                             return jtvec(jt, INT, 1L, u);
+                         }) &&
+           !jt->jerr;
+}
+
+template <typename T>
+static auto
+inplace_negate(T *u, int64_t n) {
+    std::transform(u, u + n, u, [](auto v) { return -v; });
+}
+
+template <>
+[[nodiscard]] auto
+convert<I, X>(J jt, array w, void *yv) -> bool {
+    I u[XIDIG];
+    auto const convert_one = [&](auto c) {
+        auto const b   = c == IMIN;
+        auto d         = b ? -(1 + c) : std::abs(c);
+        int64_t length = 0;
+        for (int64_t i = 0; i < XIDIG; ++i) {
+            u[i] = d % XBASE;
+            d    = d / XBASE;
+            if (u[i]) length = i;
+        }
+        ++length;
+        *u += b;
+        if (0 > c) { inplace_negate(u, XIDIG); }
+        return jtvec(jt, INT, length, u);
+    };
+    return convert<I, X>(jt, w, yv, convert_one) && !jt->jerr;
+}
+
+static auto
+jtxd1(J jt, D p, I mode) -> X {
+    PROLOG(0052);
+    D e = jttfloor(jt, p);
+    switch (mode) {
+        case XMFLR: p = e; break;
+        case XMCEIL: p = ceil(p); break;
+        case XMEXACT:
+            ASSERT(TEQ(p, e), EVDOMAIN);
+            p = e;
+            break;
+        case XMEXMT:
+            if (!TEQ(p, e)) { return jtvec(jt, INT, 0L, &iotavec[-IOTAVECBEGIN]); }
+    }
```

Comment:
```suggestion
    if (!TEQ(p, e)) return jtvec(jt, INT, 0L, &iotavec[-IOTAVECBEGIN]);
    p = mode == XMEXACT ? e : ceil(p);
```

Reply:
This refactor would invalidate the meaning of the modes (if the comments are up to date)
https://github.com/codereport/jsource/blob/5f03cf2d36db4a5be84a54bd0d0c5f86ded1ea08/jsrc/vx.h#L20-L23